docs/design-documents/promtail-scraping-objectstore.md

```diff
@@ -0,0 +1,89 @@
+# Promtail-Objectstore-scraping
+
+Promtail currently supports scraping logs from local filesystem logs, syslog and systemd/journal. This covers most of the usecases where an user would potentially store logs. In addition to this there are few other usecases where the logs can be stored in an Object store and user would like to collect and aggregate those logs in Loki. AWS loadbalancer is an one such example, Loadbalancer access logs are stored in S3. It would be good addition to Promtail to support scraping logs from Object store to cover similar user cases mentioned above. 
+
+Cortex already has the [implementation](https://github.com/cortexproject/cortex/blob/master/pkg/chunk/aws/s3_storage_client.go) for Objecstore including S3 and GCS. We can try to re-use most of it. Still, we have to develop code for `file watching` and updating the `positions.yaml` file.  Following are the few challenges which we should try to address.
+
+### There is no appending of lines to file in object store
+
+In an object store there is no appending of lines to a file in bucket. Any file modified will be newly 
+uploaded to the bucket with the same name. So, we would have to read the entire file again from the begining since we will not be aware if only new lines were appended to the modified file or the file is completely having new content
+
+### Where should we keep positions.yaml file ?
+
+Since, the files we scrape is remotely placed in a objecstore, should we keep the `positions.yaml` also in the same place? Advantage of keeping the `positions.yaml` within the objecstore will help us to recover from the state where the scraping was left when promtail is running in docker container and the container crashes. However, `user` reading the files from the Object store would also need write access to the Objecstore. Getting write access may not be possible in some cases where people doesn't want to risk modifying the files by giving write access.
+
+### How should we support reading of compressed files ?
+
+If the files in the Objectstore is stored in a compressed format, promtail should have support for reading the compressed data and we should make sure we support most of the compression algorithms like we already do for chunks(Snappy, LZ4, Gzip) 
+
+### How we should identify and watch newly added files ?
+
+To identify and to start tailing the newly added files, we should list the files in the Object store at some desired intervals. How frequently we should this? Once we add the file to `watch`, to identify any changes to the file we should frequently check the modified `datetime` of the file. The list of files might be large and each watch will be running in a `go routine` and each `go routine` making frequent `http` requests to the Obejctstore. We should make sure this doesn't affect promtail's performance.
```

Comment:
I'm not worried about goroutine. I think the frequency should be configurable.

Reply:
Just a thought that since this is a client side code we should make sure we don't consume lot of CPU cycles

Reply:
S3 is event driven and can be configured to trigger an event when updates occur.  Setting up a Lambda in combination with SNS, promtail can be notified that new content has arrived.  I’m ingesting data from S3 into Loki via promtail today using a similar architecture.  This eliminates polling for changes to S3. I am interested in your proposed feature as I would use it for much more than load balancer logs.

Reply:
Hi @mrmikemarshall 
Thanks for the inputs. We definitely want to consider other options like SQS. However, SQS doesn't support FIFO and we might get out of order entries due to this. Also, as @Punkoivan mentioned below, AWS doesn't have any plan to support FIFO

Reply:
In our preliminary version, we just want to keep things simple. Later, we can plan to support both general polling and SQS